File: frontend/console/src/api/businessDate.ts
import { apiGet, apiPost } from "./client";

export type BusinessDateStatus =
  | "open"
  | "closing"
  | "closed";

export type BusinessDate = {
  id: string;
  business_date: string;
  status: BusinessDateStatus;
  opened_at: string;
  closing_started_at: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type NightAuditReservationItem = {
  reservation_id: string;
  external_id: string | null;
  check_in: string;
  check_out: string;
  room_id: string | null;
};

export type NightAuditRoomChargeCandidate = {
  reservation_id: string;
  folio_id: string;
  service_date: string;
  amount: string;
};

export type NightAuditRoomChargeBlocker = {
  reservation_id: string;
  service_date: string;
  amount: string;
  reason: string;
};

export type NightAuditWorklist = {
  business_date: BusinessDate;
  unresolved_arrivals: NightAuditReservationItem[];
  unresolved_departures: NightAuditReservationItem[];
  room_charge_candidates: NightAuditRoomChargeCandidate[];
  room_charge_blockers: NightAuditRoomChargeBlocker[];
};

export type StartNightAuditResponse = {
  business_date: BusinessDate;
};

export type FinalizeNightAuditResponse = {
  closed_business_date: BusinessDate;
  current_business_date: BusinessDate;
};

export type PostNightAuditRoomChargesResponse = {
  posted_room_charges: NightAuditRoomChargeCandidate[];
};

export function getCurrentBusinessDate(): Promise<BusinessDate> {
  return apiGet<BusinessDate>("/business-date/current");
}

export function getNightAuditWorklist(): Promise<NightAuditWorklist> {
  return apiGet<NightAuditWorklist>("/business-date/night-audit/worklist");
}

export function startNightAudit(): Promise<StartNightAuditResponse> {
  return apiPost<StartNightAuditResponse, { reason: string }>(
    "/business-date/night-audit/start",
    { reason: "console night audit start" },
  );
}

export function postNightAuditRoomCharges(): Promise<PostNightAuditRoomChargesResponse> {
  return apiPost<PostNightAuditRoomChargesResponse, undefined>(
    "/business-date/night-audit/post-room-charges",
    undefined,
  );
}

export function finalizeNightAudit(): Promise<FinalizeNightAuditResponse> {
  return apiPost<FinalizeNightAuditResponse, { reason: string }>(
    "/business-date/night-audit/finalize",
    { reason: "console night audit finalize" },
  );
}

export function markNightAuditArrivalNoShow(
  reservationId: string,
): Promise<unknown> {
  return apiPost<unknown, undefined>(
    `/business-date/night-audit/arrivals/${reservationId}/no-show`,
    undefined,
  );
}

export function extendNightAuditDeparture(
  reservationId: string,
): Promise<unknown> {
  return apiPost<unknown, undefined>(
    `/business-date/night-audit/departures/${reservationId}/extend-stay`,
    undefined,
  );
}

File: backend/tests/common/business_date.rs
use axum::{http::StatusCode, Router};

use chrono::NaiveDate;

use serde_json::{json, Value};

use super::client::{get, post_json, response_json};

pub async fn get_current_business_date(app: &Router) -> Value {
    let response = get(app, "/business-date/current").await;

    assert_eq!(response.status(), StatusCode::OK);

    response_json(response).await
}

pub async fn current_open_business_date(app: &Router) -> NaiveDate {
    let current = get_current_business_date(app).await;

    assert_eq!(current["status"], "open");

    let business_date = current["business_date"].as_str().unwrap();

    NaiveDate::parse_from_str(business_date, "%Y-%m-%d").unwrap()
}

pub async fn start_night_audit(app: &Router) -> Value {
    let response = post_json(
        app,
        "/business-date/night-audit/start",
        &json!({ "reason": "test night audit" }),
    )
    .await;

    assert_eq!(response.status(), StatusCode::OK);

    response_json(response).await
}

pub async fn finalize_night_audit(app: &Router) -> Value {
    let response = post_json(
        app,
        "/business-date/night-audit/finalize",
        &json!({ "reason": "test night audit" }),
    )
    .await;

    assert_eq!(response.status(), StatusCode::OK);

    response_json(response).await
}

pub async fn get_night_audit_worklist(app: &Router) -> Value {
    let response = get(app, "/business-date/night-audit/worklist").await;

    assert_eq!(response.status(), StatusCode::OK);

    response_json(response).await
}

pub async fn post_night_audit_room_charges(app: &Router) -> Value {
    let response = post_json(
        app,
        "/business-date/night-audit/post-room-charges",
        &json!({}),
    )
    .await;

    assert_eq!(response.status(), StatusCode::OK);

    response_json(response).await
}

pub async fn advance_business_date(app: &Router) -> Value {
    start_night_audit(app).await;
    post_night_audit_room_charges(app).await;
    finalize_night_audit(app).await
}
